Implement `chcpu`

alxndrv opened the pull request:
This PR implements pretty much the entirety of `chcpu`.

Some things still missing:
- Tests: I'm not entirely sure how to properly test something which is so tightly coupled to system internals.
- Exit codes: `chcpu` should return a non-zero exit code in some edge cases. This implementation does check for most such cases, but pretty much always returns a success exit-code. I'm guessing that `uucore::error::UResult` can be used to solve this somehow without having to manually call `std::process::exit()`.

sylvestre commented:
in general, to test such things, you either try to mock up
and you do your best to look at the error management

sylvestre commented on src/uu/chcpu/src/chcpu.rs:
```diff
@@ -0,0 +1,379 @@
+    out
+}
+
+#[test]
```
could you please move the tests at the end of the file? thanks

sylvestre commented on src/uu/chcpu/src/chcpu.rs:
```diff
@@ -0,0 +1,379 @@
+    }
+
+    fn is_online(&self) -> bool {
+        if let Ok(state) = fs::read_to_string(self.get_path().join("online")) {
```
I think it can be simplified like

    fs::read_to_string(self.get_path().join("online"))
        .map(|state| match state.trim() {
            "0" => false,
            "1" => true,
            other => panic!("Unrecognized CPU online state: {}", other),
        })
        .unwrap_or(true) 

no ?

sylvestre commented on src/uu/chcpu/src/chcpu.rs:
```diff
@@ -0,0 +1,379 @@
+        }
+
+        let result =
+            File::create(self.get_path().join("online")).and_then(|mut f| f.write_all(b"1"));
```
looks like you are duplicating the code a bit in these 3 functions.
what about creating something like ?

fn write_cpu_state(path: PathBuf, value: &[u8], action: &str, cpu_id: u32) {
    let result = File::create(path).and_then(|mut f| f.write_all(value));
    match result {
        Ok(_) => println!("CPU {} {}", cpu_id, action),
        Err(e) => println!("CPU {} {} failed: {}", cpu_id, action, e.kind()),
    }
}

sylvestre commented on src/uu/chcpu/src/chcpu.rs:
```diff
@@ -0,0 +1,379 @@
+    }
+
+    fn enable(&self) {
+        if !self.is_hotpluggable() {
+            println!("CPU {} is not hot-pluggable", self.0);
+            return;
+        }
+
+        if self.is_online() {
+            println!("CPU {} is already enabled", self.0);
+            return;
+        }
```
you can also dedup this

alxndrv replied:
Don't think this one can really be deduped? The second check is the inverse of the one here in the case of `Cpu::disable()`, (and the messages are different too).

sylvestre commented on src/uu/chcpu/src/chcpu.rs:
```diff
@@ -0,0 +1,379 @@
+    };
+}
+
+fn enable_cpus(cpu_list: &str) {
+    let to_enable = parse_cpu_list(cpu_list).into_iter().map(Cpu);
+
+    for cpu in to_enable {
+        if cpu.ensure_exists() {
+            cpu.enable();
+        };
+    }
+}
+
+fn disable_cpus(cpu_list: &str) {
+    let to_disable = parse_cpu_list(cpu_list).into_iter().map(Cpu);
+
+    for cpu in to_disable {
+        if cpu.ensure_exists() {
+            cpu.disable();
+        }
+    }
+}
+
+fn configure_cpus(cpu_list: &str) {
+    let to_configure = parse_cpu_list(cpu_list).into_iter().map(Cpu);
+    for cpu in to_configure {
+        if cpu.ensure_exists() {
+            cpu.configure();
+        }
+    }
+}
+
+fn deconfigure_cpus(cpu_list: &str) {
+    let to_deconfigure = parse_cpu_list(cpu_list).into_iter().map(Cpu);
+    for cpu in to_deconfigure {
+        if cpu.ensure_exists() {
+            cpu.deconfigure();
+        }
+    }
+}
```
what about something like ?

```suggestion
fn process_cpus<F>(cpu_list: &str, action: F)
where
    F: Fn(&Cpu)
{
    parse_cpu_list(cpu_list)
        .into_iter()
        .map(Cpu)
        .filter(|cpu| cpu.ensure_exists())
        .for_each(|cpu| action(&cpu));
}

// Usage becomes:
fn enable_cpus(cpu_list: &str) {
    process_cpus(cpu_list, |cpu| cpu.enable());
}

fn disable_cpus(cpu_list: &str) {
    process_cpus(cpu_list, |cpu| cpu.disable());
}

fn configure_cpus(cpu_list: &str) {
    process_cpus(cpu_list, |cpu| cpu.configure());
}

fn deconfigure_cpus(cpu_list: &str) {
    process_cpus(cpu_list, |cpu| cpu.deconfigure());
}
```

alxndrv commented:
@sylvestre Thanks for the review, I'll implement your suggestions a bit later. Since we're on the topic of deduping: there's some code here that's copy-pasted from my work on `lscpu` (the `parse_cpu_list()` function). Should there be some kind of "lib" crate for common stuff like this? These kinds of CPU lists come up in a bunch of Linux utils and theres no sense in implementing it all over again every time. There's other similar examples: for example the mount-related utils all have to parse mount entries from `/proc/<pid>/mountinfo` and could benefit from having a shared lib.

sylvestre commented:
> Should there be some kind of "lib" crate for common stuff like this? 

yes, we try to publish content in uucore
https://docs.rs/uucore/latest/uucore/

alxndrv commented:
Updated the implementation to make use of `UResult`s in order to return correct non-zero exit codes in most failure scenarios. Also added some basic test scenarios and cleaned up the code a bit.

alxndrv commented:
> > Should there be some kind of "lib" crate for common stuff like this?
> 
> yes, we try to publish content in uucore https://docs.rs/uucore/latest/uucore/

In regards to this, good to know. I can look into adding the shared CPU list parsing functionality to that later, and then replace the implementations present in `util-linux` in a separate PR. Does that sound fine?

sylvestre commented:
I am really really sorry but I took this https://github.com/uutils/util-linux/pull/234 instead. it was more complete :(

sorry again :(
